Add config for custom metrics tags

vcrfxia opened the pull request:
### Description 

Adds a new config `ksql.metrics.tags.custom` which allows users to add custom tags to KSQL engine metrics. The value of the config is a string of semicolon-separated key-value pairs of tags to add. For example, passing the value `key1:value1;key2:value2` adds two tags to each metric, one with name `key1` and value `value1`, the other with name `key2` and value `value2`. The default value for the config is the empty string which results in no custom tags being added.

The config currently only applies to the new metrics format (see https://github.com/confluentinc/ksql/pull/2831), but if preferable I can add the custom tags to the old format as well.

Also unsure whether this change is worth adding to the either the changelog or the server config docs. Thoughts?

### Testing done 

Updated unit tests.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

spena commented on ksql-engine/src/main/java/io/confluent/ksql/internal/KsqlEngineMetrics.java:
```diff
@@ -315,19 +324,36 @@ private void configureNumActiveQueriesForGivenState(
+              "Invalid config value for '%s'. value: %s. reason: %s",
+              KsqlConfig.KSQL_CUSTOM_METRICS_TAGS,
+              tags,
+              e.getMessage()));
```
Wouldn't be better to catch this error early in the `KsqlContext`?

I feel we should pass the tags to the `KsqlEngineMetrics` already parsed into a Map. Also, it makes sense to have one place where the `KsqlConfig.KSQL_CUSTOM_METRICS_TAGS` is used instead of having 1 in KsqlContext to get the tags, and another in the metrics to throw the exception.

Better yet, would it make sense to have a `KsqlConfig` method that parses these key/value pairs? It might be useful in future configs with similar values. Like `Map<String, String> getKeyValues(String key)` or a different name.

agavra commented on ksql-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java:
```diff
@@ -66,14 +66,15 @@ public KsqlEngine(
       final ProcessingLogContext processingLogContext,
       final FunctionRegistry functionRegistry,
-      final String serviceId
+      final String serviceId,
```
nit: I think we should couple `serviceId` and `custommetricsTags` into a `ServiceInfo` class so that piping this type of information down will be easier in the future (also I imagine that `serviceId` might be valuable for metrics as well)

vcrfxia commented:
> Also, any reason not to use the LIST type from AbstractConfig here?

No strong reason but if we were to use the LIST type we'd still have to parse the individual tags into name-value pairs, and add in logic to throw an error if tag names are duplicated since `KafkaMetric` wants a `Map` of tags, not a `List`. Do you still think that's preferable? @rodesai If so I'm happy to make the change since I don't feel too strongly.

rodesai commented:
> Do you still think that's preferable?

Yes I think so. I think this config is pretty useful, and we should try and keep it consistent with the rest of cp. 

> Also unsure whether this change is worth adding to the either the changelog or the server config docs. Thoughts?

Yes, we should add it to both.

vcrfxia commented:
Thanks for the reviews, all! Added two commits.

First one should be uncontroversial: 

- Moved config validation out of KsqlEngineMetrics and into KsqlConfig (called from KsqlContext)
- Switched delimiter between key:value pairs from semicolon to comma, to be consistent with the delimiter used by the AbstractConfig's List type
- Added docs for the new config, and a short entry to the changelog.

The second commit introduces the ServiceInfo object @agavra suggested. Feels a little overkill at the moment but I do think it's cleaner. Happy to hear what others think, can easily back it out.

spena commented on docs/changelog.rst:
```diff
@@ -18,6 +18,9 @@ KSQL 5.4.0 includes new features, including:
   Avro record or JSON object, depending on the format in use, or as an anonymous value.
   For more information, refer to :ref:`ksql_single_field_wrapping`.
 
+* A new config `ksql.metrics.tags.custom` for adding custom tags to emitted JMX metrics.
```
is it `to emit` or `with emitted`?

vcrfxia replied:
I have a slight preference for "adding ... to" versus "adding ... with" but I don't feel too strongly.

spena commented on ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java:
```diff
@@ -718,6 +719,22 @@ public KsqlConfig overrideBreakingConfigsWithOriginalValues(final Map<String, St
+  public Map<String, String> getStringAsMap(final String key) {
+    final String value = getString(key).trim();
+    try {
+      return value.equals("")
```
Check for the NULL value as well. I think if you set a config like `config=` without a value, then the value returned will be null. I recalled seen this null value in some tests with configs.

vcrfxia replied:
Hm, I tried setting `ksql.metrics.tags.custom=` just now and got the empty string. Are you able to reproduce getting a null value?

spena replied:
Perhaps it was a different configuration that has NULL as default.

JimGalasyn commented on docs/changelog.rst:
```diff
@@ -18,6 +18,9 @@ KSQL 5.4.0 includes new features, including:
   Avro record or JSON object, depending on the format in use, or as an anonymous value.
   For more information, refer to :ref:`ksql_single_field_wrapping`.
 
+* A new config `ksql.metrics.tags.custom` for adding custom tags to emitted JMX metrics.
```
```suggestion
* A new config ``ksql.metrics.tags.custom`` for adding custom tags to emitted JMX metrics.
```

vcrfxia replied:
Thanks for the catch @JimGalasyn ! I updated an earlier line in the changelog with the same fix while I was at it.
